src/main/java/com/demo/classroom/Security/Service/JwtService.java

```
@@ -0,0 +1,134 @@
+        this.JWT_EXPIRATION = 900000; 
+        this.REFRESH_EXPIRATION_TIME=3600000;
+    }
+    
+    public String extractUsername(String token) {
+        return extractClaim(token, Claims::getSubject);
+    }
+
+    public Long extractUserId(String token) {
+        Claims claims = extractAllClaims(token);
+        return claims.get("userId", Long.class);  
+    }
+
+    public List<String> extractRoles(String token) {
+        Claims claims = extractAllClaims(token);
+        return claims.get("roles", List.class);
+    }
+
+    public <T> T extractClaim(String token, Function<Claims, T> claimsResolver) {
+        final Claims claims = extractAllClaims(token);
+        return claimsResolver.apply(claims);
+    }
+    
+    private Claims extractAllClaims(String token){
+        return Jwts
+                .parserBuilder()
+                .setSigningKey(getSignInKey())
+                .build()
+                .parseClaimsJws(token)
+                .getBody();
+    }
+
+    public String generateToken(UserDetails userDetails){
+        Map<String, Object> claims = new HashMap<>();
+        claims.put("roles", userDetails.getAuthorities().stream()
+            .map(GrantedAuthority::getAuthority)
+            .collect(Collectors.toList()));
+        return generateToken(claims, userDetails);
+    }
+
+    public String generateToken(UserDetails userDetails, Long userID ){
+        Map<String, Object> claims = new HashMap<>();
+        claims.put("roles", userDetails.getAuthorities().stream()
+            .map(GrantedAuthority::getAuthority)
+            .collect(Collectors.toList()));
+        
+        claims.put("userId", userID);
+
+        return generateToken(claims, userDetails);
+    }
+
+    
+
+    public String refreshAccessToken(String refreshToken, UserDetails userDetails) {
+        if (isTokenValid(refreshToken, userDetails)) {
+            return generateToken(userDetails);
+        }
+        return null;
```

Returning null when the token is invalid is not ideal. Instead, you could throw a custom exception (e.g., InvalidTokenException) to indicate why the token generation failed.

---

Do we want to throw exception here? if invalid it should redirect to `login` page instead of throwing exception?

---

redirecting should be handled from frontend. We just need to add exception in here

---

done